### PR TITLE
Partial model loading for low-VRAM systems

### PR DESCRIPTION
**Pull request by RyanJDick**
## Summary

This PR updates the model cache behavior with the following improvements:
- Added an option to enable partial model loading. If enabled, large models can be partially loaded onto the CUDA device and then streamed to the device during inference. This enables inference of large models on systems with limited VRAM (e.g. full FLUX dev inference on an 8GB GPU).
- Updated the RAM and VRAM cache size logic to more effectively use the available memory under the default configs.

## Configuration

The following invokeai.yml configs were added / changed:
- `ram`: Now defaults to None (unset), resulting in automatic RAM cache size management. It can still be set to override the default behavior.
- `vram`: Now defaults to None (unset), resulting in automatic VRAM cache size management. It can still be set to override the default behavior.
- `lazy_offload`: This config is deprecated and will be removed in a future migration. It no longer has any effect.
- `device_working_mem_gb`: This is the new recommended config for tuning the balance between working VRAM and cache VRAM. The default of 2gb should work for most use cases.
- `enable_partial_loading`: Defaults to False. Set it to True to get the new partial loading behavior for low-vram operation.

The recommended invokeai.yaml configuration for users wishing to opt-in to the new partial loading behaviour is:
```yaml
# Comment-out any existing ram/vram overrides:
# ram = 16
# vram = 12

# device_working_mem_gb defaults to 2.
# Uncomment and increase it if you are doing non-standard operations that cause OOM errors.
# (File a bug report if you are hitting unexpected OOM errors)
# device_working_mem_gb: 2

# Enable partial loading behaviour.
enable_partial_loading: True
```

## QA Instructions

I have tested a wide variety of models (SD1, SDXL, FLUX, SD3, LoRAs, quantized/non-quantized) with both `enable_partial_loading: True` and `enable_partial_loading: False`.

The scope of this change is large with risk of regressions on particular OSs or HW configurations. We will want to keep a close eye on this change in the RC stage for possible bugs.

## Checklist

- [ ] _The PR has a short but descriptive title, suitable for a changelog_
- [ ] _Tests added / updated (if applicable)_
- [ ] _Documentation added / updated (if applicable)_
- [ ] _Updated `What's New` copy (if doing a release after this PR)_

### CODE REVIEW

File: invokeai/app/api/routers/model_manager.py [hunk 1]
```diff
@@ -848,74 +846,6 @@ async def get_starter_models() -> StarterModelResponse:
     return StarterModelResponse(starter_models=starter_models, starter_bundles=starter_bundles)
 
 
-@model_manager_router.get(
-    "/model_cache",
-    operation_id="get_cache_size",
-    response_model=float,
-    summary="Get maximum size of model manager RAM or VRAM cache.",
-)
-async def get_cache_size(cache_type: CacheType = Query(description="The cache type", default=CacheType.RAM)) -> float:
```
Comment (RyanJDick) [inline invokeai/app/api/routers/model_manager.py:857]: @psychedelicious AFAICT, the `get_cache_size(...)` and `set_cache_size(...)` endpoints weren't being called anywhere in the frontend. Is that accurate? With the changes to how the cache size is managed, it was easiest to just rip these out

Comment (psychedelicious) [reply invokeai/app/api/routers/model_manager.py:857]: Yep that's correct.

**Review by hipsterusername — APPROVED** [review]

**Comment by psychedelicious** [comment]
Running into a few issues.

Starting with only this relevant setting in `invokeai.yaml`:
```yaml
enable_partial_loading: true
```

Restarted the app after each error.

I tried FLUX w/ IP Adapter and got an error: `RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument mat2 in method wrapper_CUDA_addmm)`

<details>

<summary>full error & context</summary>

Models:
image
image

```sh
Loading checkpoint shards: 100%|███████████████████████████████| 2/2 [00:00<00:00,  6.04it/s]
[2024-12-19 16:08:32,570]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '0d07998e-3115-44f2-bfec-5c901cef0b2c:text_encoder_2' (T5EncoderModel) onto cuda device in 4.44s. Total model size: 9334.39MB, VRAM: 9334.39MB (100.0%)
[2024-12-19 16:08:32,745]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '0d07998e-3115-44f2-bfec-5c901cef0b2c:tokenizer_2' (T5Tokenizer) onto cuda device in 0.00s. Total model size: 0.03MB, VRAM: 0.00MB (0.0%)
[2024-12-19 16:08:33,428]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '54a9ad60-0a74-45e4-a13d-37179ab87b45:text_encoder' (CLIPTextModel) onto cuda device in 0.04s. Total model size: 469.44MB, VRAM: 469.44MB (100.0%)
[2024-12-19 16:08:33,503]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '54a9ad60-0a74-45e4-a13d-37179ab87b45:tokenizer' (CLIPTokenizer) onto cuda device in 0.00s. Total model size: 0.00MB, VRAM: 0.00MB (0.0%)
[2024-12-19 16:08:35,182]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '65efbf64-9ea9-4b1d-93c4-e70db9e724d7' (CLIPVisionModelWithProjection) onto cuda device in 0.06s. Total model size: 579.77MB, VRAM: 579.77MB (100.0%)
[2024-12-19 16:08:45,329]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '046c8125-dfc8-4151-be26-f1c5b459bf29:transformer' (Flux) onto cuda device in 9.88s. Total model size: 22700.13MB, VRAM: 21476.13MB (94.6%)
[2024-12-19 16:08:45,371]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model 'ad993f2e-eb70-49d8-b4bb-9ce6bac3d74f' (XlabsIpAdapterFlux) onto cuda device in 0.01s. Total model size: 936.27MB, VRAM: 0.27MB (0.0%)
[2024-12-19 16:08:45,393]::[InvokeAI]::ERROR --> Error while invoking session c1b99b8d-79c6-4aa7-92d6-6a1f717956b7, invocation 01b50de3-ff97-4d72-ab6f-f5be43ca3614 (flux_denoise): Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument mat2 in method wrapper_CUDA_addmm)
[2024-12-19 16:08:45,393]::[InvokeAI]::ERROR --> Traceback (most recent call last):
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/session_processor/session_processor_default.py", line 129, in run_node
    output = invocation.invoke_internal(context=context, services=self._services)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/baseinvocation.py", line 300, in invoke_internal
    output = self.invoke(context)
  File "/home/bat/Documents/Code/InvokeAI/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/flux_denoise.py", line 144, in invoke
    latents = self._run_diffusion(context)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/flux_denoise.py", line 332, in _run_diffusion
    pos_ip_adapter_extensions, neg_ip_adapter_extensions = self._prep_ip_adapter_extensions(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/flux_denoise.py", line 700, in _prep_ip_adapter_extensions
    ip_adapter_extension.run_image_proj(dtype=dtype)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/flux/extensions/xlabs_ip_adapter_extension.py", line 54, in run_image_proj
    self._image_proj = self._model.image_proj(image_prompt_clip_embed)
  File "/home/bat/Documents/Code/InvokeAI/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/ip_adapter/ip_adapter.py", line 59, in forward
    clip_extra_context_tokens = self.proj(embeds).reshape(
  File "/home/bat/Documents/Code/InvokeAI/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/.venv/lib/python3.10/site-packages/torch/nn/modules/linear.py", line 117, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument mat2 in method wrapper_CUDA_addmm)

[2024-12-19 16:08:45,448]::[InvokeAI]::INFO --> Graph stats: c1b99b8d-79c6-4aa7-92d6-6a1f717956b7
                          Node   Calls   Seconds  VRAM Used
             flux_model_loader       1    0.024s     0.000G
             flux_text_encoder       1    7.783s     9.347G
                       collect       2    0.001s     9.343G
                 core_metadata       1    0.001s     9.343G
               flux_ip_adapter       1    0.002s     9.343G
                  flux_denoise       1   11.830s    20.995G
TOTAL GRAPH EXECUTION TIME:  19.640s
TOTAL GRAPH WALL TIME:  19.648s
RAM used by InvokeAI process: 22.89G (+22.031G)
RAM used to load models: 33.22G
VRAM in use: 20.986G
RAM cache statistics:
   Model cache hits: 7
   Model cache misses: 7
   Models cached: 3
   Models cleared from cache: 3
   Cache high water mark: 23.65G
```

</details>

I removed the Global Reference (IP Adapter) layer and tried FLUX w/ the ControlNet Union Pro model. Got a warning `[MODEL CACHE] Insufficient GPU memory to load model. Aborting` and then an `AssertionError`.

<details>

<summary>full error & context</summary>

```sh
[2024-12-19 16:20:39,308]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '9e901924-ff8c-4de8-8f25-931245b0b0e1:vae' (AutoEncoder) onto cuda device in 0.02s. Total model size: 159.87MB, VRAM: 159.87MB (100.0%)
[2024-12-19 16:20:40,239]::[ModelManagerService]::WARNING --> [MODEL CACHE] Insufficient GPU memory to load model. Aborting
[2024-12-19 16:20:40,240]::[ModelManagerService]::WARNING --> [MODEL CACHE] Insufficient GPU memory to load model. Aborting
[2024-12-19 16:20:40,245]::[InvokeAI]::ERROR --> Error while invoking session 46c5ab42-8a2f-47c4-9f0b-adccdebdce6d, invocation b617ff7e-a6f4-49d6-895a-5afa5894370f (flux_denoise): 
[2024-12-19 16:20:40,245]::[InvokeAI]::ERROR --> Traceback (most recent call last):
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_manager/load/model_cache/model_cache.py", line 213, in lock
    self._load_locked_model(cache_entry, working_mem_bytes)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_manager/load/model_cache/model_cache.py", line 265, in _load_locked_model
    model_bytes_loaded = self._move_model_to_vram(cache_entry, vram_available)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_manager/load/model_cache/model_cache.py", line 284, in _move_model_to_vram
    return cache_entry.cached_model.partial_load_to_vram(vram_available)
  File "/home/bat/Documents/Code/InvokeAI/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_manager/load/model_cache/cached_model/cached_model_with_partial_load.py", line 97, in partial_load_to_vram
    cur_state_dict[key] = param.to(self._compute_device, copy=True)
torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 108.00 MiB. GPU 0 has a total capacity of 23.63 GiB of which 20.44 MiB is free. Including non-PyTorch memory, this process has 23.42 GiB memory in use. Of the allocated memory 22.95 GiB is allocated by PyTorch, and 12.01 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/session_processor/session_processor_default.py", line 129, in run_node
    output = invocation.invoke_internal(context=context, services=self._services)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/baseinvocation.py", line 300, in invoke_internal
    output = self.invoke(context)
  File "/home/bat/Documents/Code/InvokeAI/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/flux_denoise.py", line 144, in invoke
    latents = self._run_diffusion(context)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/flux_denoise.py", line 289, in _run_diffusion
    controlnet_extensions = self._prep_controlnet_extensions(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/flux_denoise.py", line 553, in _prep_controlnet_extensions
    model = exit_stack.enter_context(controlnet_model)
  File "/usr/lib/python3.10/contextlib.py", line 492, in enter_context
    result = _cm_type.__enter__(cm)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_manager/load/load_base.py", line 60, in __enter__
    self._cache.lock(self._cache_record.key, None)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_manager/load/model_cache/model_cache.py", line 219, in lock
    cache_entry.unlock()
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_manager/load/model_cache/cache_record.py", line 27, in unlock
    assert self._locks >= 0
AssertionError
```

</details>


Ok, maybe I need to set `device_working_mem_gb: 6.0` (was previously unset). Now I get `KeyError: '046c8125-dfc8-4151-be26-f1c5b459bf29:transformer'`.

<details>

<summary>full error & context</summary>

```sh
Loading checkpoint shards: 100%|███████████████████████████████| 2/2 [00:00<00:00,  6.13it/s]
[2024-12-19 16:21:31,326]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '0d07998e-3115-44f2-bfec-5c901cef0b2c:text_encoder_2' (T5EncoderModel) onto cuda device in 4.27s. Total model size: 9334.39MB, VRAM: 9334.39MB (100.0%)
[2024-12-19 16:21:31,503]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '0d07998e-3115-44f2-bfec-5c901cef0b2c:tokenizer_2' (T5Tokenizer) onto cuda device in 0.00s. Total model size: 0.03MB, VRAM: 0.00MB (0.0%)
[2024-12-19 16:21:32,167]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '54a9ad60-0a74-45e4-a13d-37179ab87b45:text_encoder' (CLIPTextModel) onto cuda device in 0.04s. Total model size: 469.44MB, VRAM: 469.44MB (100.0%)
[2024-12-19 16:21:32,239]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '54a9ad60-0a74-45e4-a13d-37179ab87b45:tokenizer' (CLIPTokenizer) onto cuda device in 0.00s. Total model size: 0.00MB, VRAM: 0.00MB (0.0%)
[2024-12-19 16:21:33,371]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '9e901924-ff8c-4de8-8f25-931245b0b0e1:vae' (AutoEncoder) onto cuda device in 0.02s. Total model size: 159.87MB, VRAM: 159.87MB (100.0%)
[2024-12-19 16:21:35,146]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model 'd830a8f2-0db8-4504-aad9-ebf7b6930560' (InstantXControlNetFlux) onto cuda device in 1.59s. Total model size: 6297.98MB, VRAM: 6297.98MB (100.0%)
[2024-12-19 16:21:35,150]::[InvokeAI]::ERROR --> Error while invoking session e1fc9e22-df57-4fd2-8687-269b290f9da6, invocation fa2a3d36-0e7f-4bdf-9302-d8750b2d7752 (flux_denoise): '046c8125-dfc8-4151-be26-f1c5b459bf29:transformer'
[2024-12-19 16:21:35,150]::[InvokeAI]::ERROR --> Traceback (most recent call last):
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/session_processor/session_processor_default.py", line 129, in run_node
    output = invocation.invoke_internal(context=context, services=self._services)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/baseinvocation.py", line 300, in invoke_internal
    output = self.invoke(context)
  File "/home/bat/Documents/Code/InvokeAI/.venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/flux_denoise.py", line 144, in invoke
    latents = self._run_diffusion(context)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/flux_denoise.py", line 299, in _run_diffusion
    (cached_weights, transformer) = exit_stack.enter_context(transformer_info.model_on_device())
  File "/usr/lib/python3.10/contextlib.py", line 492, in enter_context
    result = _cm_type.__enter__(cm)
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_manager/load/load_base.py", line 75, in model_on_device
    self._cache.lock(self._cache_record.key, working_mem_bytes)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_manager/load/model_cache/model_cache.py", line 203, in lock
    cache_entry = self._cached_models[key]
KeyError: '046c8125-dfc8-4151-be26-f1c5b459bf29:transformer'

[2024-12-19 16:21:35,200]::[InvokeAI]::INFO --> Graph stats: e1fc9e22-df57-4fd2-8687-269b290f9da6
                          Node   Calls   Seconds  VRAM Used
             flux_model_loader       1    0.017s     0.000G
             flux_text_encoder       1    7.593s     9.347G
                       collect       2    0.001s     9.343G
                 core_metadata       1    0.001s     9.343G
               flux_controlnet       1    0.000s     9.343G
                  flux_denoise       1    2.855s     9.352G
TOTAL GRAPH EXECUTION TIME:  10.467s
TOTAL GRAPH WALL TIME:  10.473s
RAM used by InvokeAI process: 9.64G (+8.784G)
RAM used to load models: 38.05G
VRAM in use: 6.322G
RAM cache statistics:
   Model cache hits: 7
   Model cache misses: 7
   Models cached: 2
   Models cleared from cache: 1
   Cache high water mark: 22.17G
```

</details>

Tried with it set to 9.0, same exact error.

Just trying FLUX without any extras, full model. With 1536x1536 and `device_working_mem_gb` unset, I get a OOM during VAE decode. Setting `device_working_mem_gb` to a larger value allows it to complete.

File: invokeai/app/invocations/latents_to_image.py [hunk 2]
```diff
@@ -59,7 +59,13 @@ def invoke(self, context: InvocationContext) -> ImageOutput:
 
         vae_info = context.models.load(self.vae.vae)
         assert isinstance(vae_info.model, (AutoencoderKL, AutoencoderTiny))
-        with SeamlessExt.static_patch_model(vae_info.model, self.vae.seamless_axes), vae_info as vae:
+        # Reserve 6GB of VRAM for the VAE.
+        # Experimentally, this was found to be sufficient for decoding a 1024x1024 image.
+        # TODO(ryand): Set the requested working memory dynamically based on the image size (and self.fp32).
```
Comment (psychedelicious) [inline invokeai/app/invocations/latents_to_image.py:64]: This feels risky. What happens if we want to decode a larger image?

File: invokeai/app/services/config/config_default.py
Comment (psychedelicious) [inline invokeai/app/services/config/config_default.py:1]: It looks like this PR changes caching behaviour when `ram` and `vram` are unset, even if the user does not opt-in to the new caching with `enable_partial_loading: true`.

Am I understanding correctly, that if the user had manually set the `ram` and `vram`, that there should be no change to behaviour with this PR?

**Comment by psychedelicious** [comment]
It seems the handling of OOMs has changed.

I expected `OutOfMemoryError`s to bubble up the stack and be recorded as the error type for the node, but we get the `AssertionError` instead.  This occurs when we OOM regardless of the `enable_partial_loading` setting. The user-facing result is that we generic `Server Error` toasts instead of more descriptive toasts saying an OOM occurred. 

---

Looks like we unlock one too many times when a move to VRAM fails during a call to `ModelCache.lock()` - once in the exception handler in `_move_model_to_vram()`, then again in the top-level exception handler for `ModelCache.lock()`.

We also get two warnings - one at each of these two levels.

I'll defer to you to fix it - but if I do this, at least the OOMs bubble up correctly. 

```diff
diff --git a/invokeai/backend/model_manager/load/model_cache/model_cache.py b/invokeai/backend/model_manager/load/model_cache/model_cache.py
index 760ec5f348..d4f552e094 100644
--- a/invokeai/backend/model_manager/load/model_cache/model_cache.py
+++ b/invokeai/backend/model_manager/load/model_cache/model_cache.py
@@ -209,18 +209,8 @@ class ModelCache:
             # Models don't need to be loaded into VRAM if we're running on CPU.
             return
 
-        try:
-            self._load_locked_model(cache_entry, working_mem_bytes)
-            self._logger.debug(
-                f"Finished locking model {key} (Type: {cache_entry.cached_model.model.__class__.__name__})"
-            )
-        except torch.cuda.OutOfMemoryError:
-            self._logger.warning("Insufficient GPU memory to load model. Aborting")
-            cache_entry.unlock()
-            raise
-        except Exception:
-            cache_entry.unlock()
-            raise
+        self._load_locked_model(cache_entry, working_mem_bytes)
+        self._logger.debug(f"Finished locking model {key} (Type: {cache_entry.cached_model.model.__class__.__name__})")
 
         self._log_cache_state()

```

File: invokeai/app/services/config/config_default.py
Comment (RyanJDick) [reply invokeai/app/services/config/config_default.py:1]: Correct. The thinking is that the old defaults for ram/vram do not make very effective use of the available RAM / VRAM, so we want to give a better default experience. If ram and vram are overridden than the behavior should be very close to the old behavior (but I have not gone to the trouble of trying to guarantee that it's strictly identical)

**Comment by RyanJDick** [comment]
Some notes after reviewing the issues that @psychedelicious found:

> I tried FLUX w/ IP Adapter and got an error: RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument mat2 in method wrapper_CUDA_addmm)

The root cause is that the IP-Adapter model has multiple entrypoints. Currently, auto-device-casting only works when a model is called via its `torch.nn.module.forward(...)` method. In the future, we should be able to improve the autocast context management to support this case, but for now I have added `XlabsIpAdapterFlux` to a list of models that opt-out of partial loading (in [7fddb06](https://github.com/invoke-ai/InvokeAI/pull/7460/commits/7fddb06dc4226538dd16451b03e13465ceb1df30)). There is a risk that other models have a similar problem, but we just haven't encountered it yet.

---

Related to the other issues, the following fixes are needed:
- [ ] The working memory required for VAE decode needs to be calculated dynamically based on the size of image being decoded.
- [ ] Fix up the logic for handling exceptions during model load / unload. The model should be kicked out of the cache, and we should make sure that all lock tracking gets updated correctly.
- [ ] I'm not yet sure what is causing the OOM error during partial load of the ControlNet Union Pro model. I need to investigate this some more. @psychedelicious, if you are able to reproduce it with debug logs turned on, that would be helpful.
- [ ] When loading a model that does not support partial loading, locked models that _do_ support partial loading should be partially unloaded to make room for it. We don't currently do this.
- [ ] There is a bug in flux_denoise where we load a FLUX model, don't lock it, load other models, and then expect the FLUX model to still be in the cache. There are probably other places in our code base where we do this, since the model loading APIs don't do a good job of discouraging it. This error was possible before, but the prevalence seems to have gone up - I need to think some more about why that is.

**Comment by psychedelicious** [comment]
Marked as draft to prevent accidental merge.

**Comment by RyanJDick** [comment]
Closing in favor of smaller PRs that will be easier to test.

**Pull request closed without merging**